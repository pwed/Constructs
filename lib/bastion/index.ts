export {BastionPermissionSet, BastionPermissionSetProps} from "./lib/sso-permission-set"
export {WindowsBastion, WindowsBastionProps} from "./lib/windows-bastion"
export {ScheduleShutdown, ScheduleShutdownProps} from "./lib/schedule-shutdown"
