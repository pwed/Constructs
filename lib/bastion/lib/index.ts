exports.WindowsBastion = require('./windows-bastion');
