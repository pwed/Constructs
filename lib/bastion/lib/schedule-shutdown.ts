import { aws_iam, aws_ssm, Stack, Tag } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Constants } from './constants';

export interface ScheduleShutdownProps {
  securityTag?: Tag;
  timezone?: string;
  shutdownSchedule?: string;
}

export class ScheduleShutdown extends Construct {
  constructor(scope: Construct, id: string, props?: ScheduleShutdownProps) {
    super(scope, id);

    const securityTag = props?.securityTag
      ? props.securityTag
      : Constants.securityTag;

    const maintanenceWindow = new aws_ssm.CfnMaintenanceWindow(
      this,
      'BastionMaintanenceWindow',
      {
        allowUnassociatedTargets: false,
        cutoff: 0,
        duration: 1,
        name: Stack.of(scope).stackName + '-Maintanence-Window',
        schedule: props?.shutdownSchedule
          ? props.shutdownSchedule
          : 'cron(0 2 ? * * *)',
        scheduleTimezone: props?.timezone
          ? props.timezone
          : 'Australia/Melbourne',
      }
    );

    const maintanenceTarget = new aws_ssm.CfnMaintenanceWindowTarget(
      this,
      'BastionMaintanenceTarget',
      {
        resourceType: 'INSTANCE',
        targets: [
          {
            key: `tag:${securityTag.Key}`,
            values: [securityTag.Value],
          },
        ],
        windowId: maintanenceWindow.ref,
      }
    );

    const taskRole = new aws_iam.Role(this, 'AutomationRole', {
      inlinePolicies: {
        ec2Stop: new aws_iam.PolicyDocument({
          statements: [
            new aws_iam.PolicyStatement({
              effect: aws_iam.Effect.ALLOW,
              resources: ['*'],
              actions: ['ec2:StopInstances'],
              conditions: {
                StringEquals: JSON.parse(
                  `{"aws:ResourceTag/${securityTag.Key}": "${securityTag.Value}"}`
                ),
              },
            }),
            new aws_iam.PolicyStatement({
              effect: aws_iam.Effect.ALLOW,
              resources: ['*'],
              actions: ['ec2:DescribeInstances', 'ec2:DescribeInstanceStatus'],
            }),
          ],
        }),
      },
      assumedBy: new aws_iam.ServicePrincipal('ssm.amazonaws.com'),
    });

    new aws_ssm.CfnMaintenanceWindowTask(this, 'BastionStop', {
      priority: 1,
      taskArn: 'AWS-StopEC2Instance',
      taskType: 'AUTOMATION',
      windowId: maintanenceWindow.ref,
      taskInvocationParameters: {
        maintenanceWindowAutomationParameters: {
          documentVersion: '1',
          parameters: {
            InstanceId: ['{{RESOURCE_ID}}'],
            AutomationAssumeRole: [taskRole.roleArn],
          },
        },
      },
      maxErrors: '1',
      maxConcurrency: '1',
      targets: [
        {
          key: 'WindowTargetIds',
          values: [maintanenceTarget.ref],
        },
      ],
    });
  }
}
