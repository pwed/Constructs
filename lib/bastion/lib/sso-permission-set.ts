import { aws_iam, aws_sso, Tag } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Constants } from './constants';

interface BastionPermissionSetProps {
  ssoInstanceArn: string;
  permissionSetName: string;
  sessionDuration?: string;
  securityTag?: Tag;
}

export class BastionPermissionSet extends Construct {
  private policy: aws_iam.PolicyDocument;
  private permissionSet: aws_sso.CfnPermissionSet;
  private ssoInstanceArn: string;
  securityTag: Tag;
  constructor(scope: Construct, id: string, props: BastionPermissionSetProps) {
    super(scope, id);

    this.securityTag = props.securityTag
      ? props.securityTag
      : Constants.securityTag;

    this.ssoInstanceArn = props.ssoInstanceArn;

    this.policy = new aws_iam.PolicyDocument({
      statements: [
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: [
            'cloudwatch:DescribeAlarms',
            'ec2:DescribeInstances',
            'ec2:DescribeInstanceStatus',
            'ec2:DescribeSecurityGroup*',
            'identitystore:DescribeUser',
            'ssm-guiconnect:CancelConnection',
            'ssm-guiconnect:GetConnection',
            'ssm-guiconnect:StartConnection',
            'ssm:DescribeInstance*',
            'ssm:GetCommandInvocation',
            'ssm:GetInventorySchema',
            'sso:ListDirectoryAssociations*',
            'rds:Describe*',
            'secretsmanager:ListSecrets',
            'kms:ListAliases',
          ],
          resources: ['*'],
        }),
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: [
            'ec2:GetPasswordData',
            'ec2:StartInstances',
            'ssm:GetConnectionStatus',
            'ssm:SendCommand',
            'ssm:StartSession',
          ],
          resources: ['arn:aws:ec2:*:*:instance/*'],
          conditions: {
            StringEquals: JSON.parse(
              `{"aws:ResourceTag/${this.securityTag.key}": "${this.securityTag.value}"}`
            ),
          },
        }),
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: [
            'secretsmanager:DescribeSecret',
            'secretsmanager:GetSecretValue',
          ],
          resources: ['*'],
          conditions: {
            StringEquals: JSON.parse(
              `{"aws:ResourceTag/${this.securityTag.key}": "${this.securityTag.value}"}`
            ),
          },
        }),
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: ['ssm:TerminateSession'],
          resources: ['*'],
          conditions: {
            StringLike: {
              'ssm:resourceTag/aws:ssmmessages:session-id': '${aws:userName}',
            },
          },
        }),
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: ['ssm:GetDocument'],
          resources: [
            'arn:aws:ssm:*:*:document/AWS-StartPortForwardingSession',
            'arn:aws:ssm:*:*:document/SSM-SessionManagerRunShell',
          ],
        }),
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.ALLOW,
          actions: ['ssm:SendCommand', 'ssm:StartSession'],
          resources: [
            'arn:aws:ssm:*:*:document/AWS-StartPortForwardingSession',
            'arn:aws:ssm:*:*:document/AWSSSO-CreateSSOUser',
            'arn:aws:ssm:*:*:managed-instance/*',
          ],
          conditions: {
            BoolIfExists: {
              'ssm:SessionDocumentAccessCheck': 'true',
            },
          },
        }),
      ],
    });

    this.permissionSet = new aws_sso.CfnPermissionSet(
      this,
      'BastionAccessPermissionSet',
      {
        instanceArn: props.ssoInstanceArn,
        inlinePolicy: this.policy,
        name: props.permissionSetName
          ? props.permissionSetName
          : 'BastionAccess',
        sessionDuration: props.sessionDuration ? props.sessionDuration : 'PT8H',
        relayStateType:
          'https://console.aws.amazon.com/systems-manager/managed-instances/rdp-connect',
      }
    );
  }

  assign(accountId: string, principalId: string, principalType: string): void {
    new aws_sso.CfnAssignment(
      this,
      `Account${accountId}${principalType}${principalId}`,
      {
        permissionSetArn: this.permissionSet
          .getAtt('PermissionSetArn')
          .toString(),
        principalType,
        principalId,
        instanceArn: this.ssoInstanceArn,
        targetType: 'AWS_ACCOUNT',
        targetId: accountId,
      }
    );
  }
}
