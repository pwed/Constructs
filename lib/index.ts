exports.static_site = require('./static-site');
