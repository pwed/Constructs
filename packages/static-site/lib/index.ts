import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export interface StaticSiteProps {
  readonly name: string;
}

export class StaticSite extends Construct {

  constructor(scope: Construct, id: string, props: StaticSiteProps) {
    super(scope, id);

    
    new Bucket(this, "bucket", {
      bucketName: props.name,
    })
  }
}
