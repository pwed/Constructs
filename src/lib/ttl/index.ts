import {
  Duration,
  aws_lambda,
  aws_events,
  aws_events_targets,
  aws_iam,
  Fn,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface TtlProps {
  readonly ttl: Duration;
  readonly pollInterval?: Duration;
}

export class Ttl extends Construct {
  constructor(scope: Construct, id: string, props: TtlProps) {
    super(scope, id);

    const ttl = props.ttl;
    const pollInterval = props.pollInterval
      ? props.pollInterval
      : Duration.hours(1);

    const ttlLambda = new aws_lambda.Function(this, 'TtlFunction', {
      runtime: aws_lambda.Runtime.PYTHON_3_9,
      code: aws_lambda.Code.fromInline(`#!/bin/env python
import boto3
import os
import time
import logging

log = logging.Logger("logger")
log.setLevel("INFO")

def handler(event, context):
    print(event)

    stack_name = os.environ['STACK_NAME']
    ttl = int(os.environ['TTL'])
    cfn = boto3.client('cloudformation')

    print(
        f"""
        StackName: {stack_name}
        TTL:       {ttl}
        """
    )

    stack = cfn.describe_stacks(StackName=stack_name)
    creation_time = stack['Stacks'][0]['CreationTime']
    creation_time_epoch = int(time.mktime(creation_time.timetuple()))

    now = time.time()
    delta = now - creation_time_epoch
    print(
        f"""
        Now:           {now}
        Creation Time: {creation_time}
        Delta:         {delta}
        """
    )
    if delta > ttl:
        print(f"deleting stack: {stack_name}")
        cfn.delete_stack(StackName=stack_name)
        return

    print(f"not deleting stack yet, {delta} seconds remaining")
    return`),
      handler: 'index.handler',
      environment: {
        TTL: ttl.toSeconds().toString(),
        STACK_NAME: Fn.ref('AWS::StackName'),
      },
      role: new aws_iam.Role(this, 'TtlRole', {
        assumedBy: new aws_iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          aws_iam.ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess'),
        ],
      }),
    });

    new aws_events.Rule(this, 'IntercalSchedule', {
      schedule: aws_events.Schedule.rate(pollInterval),
      targets: [new aws_events_targets.LambdaFunction(ttlLambda)],
    });
  }
}
