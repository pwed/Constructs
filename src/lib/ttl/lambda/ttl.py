#!/bin/env python
import boto3
import os
import time
import logging

log = logging.Logger("logger")
log.setLevel("INFO")

def handler(event, context):
    print(event)

    stack_name = os.environ['STACK_NAME']
    ttl = int(os.environ['TTL'])
    cfn = boto3.client('cloudformation')

    print(
        f"""
        StackName: {stack_name}
        TTL:       {ttl}
        """
    )

    stack = cfn.describe_stacks(StackName=stack_name)
    creation_time = stack['Stacks'][0]['CreationTime']
    creation_time_epoch = int(time.mktime(creation_time.timetuple()))

    now = time.time()
    delta = now - creation_time_epoch
    print(
        f"""
        Now:           {now}
        Creation Time: {creation_time}
        Delta:         {delta}
        """
    )
    if delta > ttl:
        print(f"deleting stack: {stack_name}")
        cfn.delete_stack(StackName=stack_name)
        return

    print(f"not deleting stack yet, {delta} seconds remaining")
    return