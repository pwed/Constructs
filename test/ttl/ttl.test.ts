import * as cdk from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as Ttl from '../../src/lib/ttl';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'TestStack');

new Ttl.Ttl(stack, 'MyTestConstruct', {
  ttl: Duration.hours(4),
});
const template = Template.fromStack(stack);

test('Has Lambda', () => {
  template.hasResource('AWS::Lambda::Function', {});
});

test('Has Role', () => {
  template.hasResource('AWS::IAM::Role', {});
});

test('Matches Snapshot', () => {
  expect(template).toMatchSnapshot();
});
