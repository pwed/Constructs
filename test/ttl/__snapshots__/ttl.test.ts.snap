// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eb5b005c858404ea0c8f68098ed5dcdf5340e02461f149751d10f59c210d5ef8.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyTestConstructIntervalSchedule13244702": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 hour)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "MyTestConstructTtlFunction4D5B5376",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyTestConstructIntervalScheduleAllowEventRuleTestStackMyTestConstructTtlFunction4874FE208BC79123": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructTtlFunction4D5B5376",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructIntervalSchedule13244702",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyTestConstructTtlFunction4D5B5376": Object {
      "DependsOn": Array [
        "MyTestConstructTtlRole7D1EB42B",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "#!/bin/env python
import boto3
import os
import time
import logging

log = logging.Logger(\\"logger\\")
log.setLevel(\\"INFO\\")

def handler(event, context):
    print(event)

    stack_name = os.environ['STACK_NAME']
    ttl = int(os.environ['TTL'])
    cfn = boto3.client('cloudformation')

    print(
        f\\"\\"\\"
        StackName: {stack_name}
        TTL:       {ttl}
        \\"\\"\\"
    )

    stack = cfn.describe_stacks(StackName=stack_name)
    creation_time = stack['Stacks'][0]['CreationTime']
    creation_time_epoch = int(time.mktime(creation_time.timetuple()))

    now = time.time()
    delta = now - creation_time_epoch
    print(
        f\\"\\"\\"
        Now:           {now}
        Creation Time: {creation_time}
        Delta:         {delta}
        \\"\\"\\"
    )
    if delta > ttl:
        print(f\\"deleting stack: {stack_name}\\")
        cfn.delete_stack(StackName=stack_name)
        return

    print(f\\"not deleting stack yet, {delta} seconds remaining\\")
    return",
        },
        "Environment": Object {
          "Variables": Object {
            "STACK_NAME": Object {
              "Ref": "AWS::StackName",
            },
            "TTL": "14400",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructTtlRole7D1EB42B",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructTtlFunctionLogRetention3ECCB05B": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "MyTestConstructTtlFunction4D5B5376",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "MyTestConstructTtlRole7D1EB42B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
